core/src/km_core_state_context_set_if_needed.cpp

```diff
@@ -175,73 +176,84 @@ context_previous_char(
   assert(new_context_string != nullptr);
   assert(context != nullptr);
   if (new_context_string == nullptr || context == nullptr) {
     return change_type;
   }
 
-  std::unique_ptr<km_core_cp[]> context_string(get_context_as_string(context));
-  if (context_string[0] == '\0') {
-    // If the app_context is "empty" then it needs updating
-    return change_type;
+  km_core_context_item *new_context_items = nullptr;
+  km_core_context_item *context_items = nullptr;
+  km_core_context_item const *context_p = nullptr, *new_context_p = nullptr;
+  uint8_t n = 0;
+
+  if(context_items_from_utf16(new_context_string, &new_context_items) != KM_CORE_STATUS_OK) {
+    goto is_context_unchanged_cleanup;
   }
 
-  km_core_context_item *context_items;
   if (km_core_context_get(context, &context_items) != KM_CORE_STATUS_OK) {
-    return change_type;
+    goto is_context_unchanged_cleanup;
+  }
+
+  if(context_items->type == KM_CORE_CT_END) {
+    // If the app_context is "empty" then it needs updating
+    goto is_context_unchanged_cleanup;
   }
 
   // move to the end of the new and internal app context
-  km_core_cp const *new_context_p = new_context_string;
-  while (*new_context_p) {
+  new_context_p = new_context_items;
+  while (new_context_p->type != KM_CORE_CT_END) {
     new_context_p++;
   }
 
-  km_core_context_item const *context_p = context_items;
+  context_p = context_items;
   while (context_p->type != KM_CORE_CT_END) {
     context_p++;
   }
 
+  // Move to last character for each pointer
+  new_context_p = context_previous_char(new_context_p, new_context_items);
+  context_p = context_previous_char(context_p, context_items);
+
   // we need to compare from the end of the cached context
-  for (; new_context_p >= new_context_string && context_p != NULL;
-        new_context_p--, context_p = context_previous_char(context_p, context_items)) {
-    if (*new_context_p != context_p->character && context_p->type != KM_CORE_CT_END) {
+  for (; new_context_p != NULL && context_p != NULL;
+        new_context_p = context_previous_char(new_context_p, new_context_items),
+        context_p = context_previous_char(context_p, context_items)) {
+    if (new_context_p->character != context_p->character) {
       // The cached context doesn't match the application context
-      km_core_context_items_dispose(context_items);
-      return change_type;
+      goto is_context_unchanged_cleanup;
     }
   }
 
-  if (context_p == NULL && new_context_p < new_context_string) {
+  if (context_p == NULL && new_context_p == NULL) {
     // new and internal app contexts are identical
     change_type.type = CONTEXT_UNCHANGED;
-    km_core_context_items_dispose(context_items);
-    return change_type;
+    goto is_context_unchanged_cleanup;
   }
 
   if (context_p >= context_items) {
-    uint8_t n = 1;
-    for (; context_p > context_items; context_p--) {
+    for (n = 1; context_p > context_items; context_p--) {
```

Comment:
I note that here we are just happy with the decrement, where as earlier it used `context_previous_char` which skips markers. For  `new_context_items` it doesn't matter as it is created using `context_items_from_utf16` ensuring markers are already stripped.

Reply:
Looking at this again, this is actually what needs to be done. It needs to count both the markers and the characters in the count difference as it will then use this count to shrink the cached_context.  It was correct not use `context_previous_char`

Reply:
Hopefully reflected in the function description now